File: back-end/app/templates.py
from datetime import timezone
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from typing import Optional
from datetime import datetime
import uuid

from .database import get_db
from .models import Template, Notification, User, generate_template_id
from .auth import get_current_user, get_current_user_required, get_current_admin_user
from .schemas import (
    TemplateCreate, TemplateUpdate, TemplateResponse,
    TemplateListResponse
)


router = APIRouter(tags=["templates"])

# Configuration
OFFICIAL_REPO_PREFIX = "https://github.com/scalebox/official-templates/"
PRIVATE_REPO_PREFIX = "https://github.com/scalebox/private-templates/"


def generate_repository_url(template_id: str, is_official: bool = False) -> str:
    """Generate repository URL based on template type"""
    if is_official:
        return f"{OFFICIAL_REPO_PREFIX}{template_id}"
    else:
        return f"{PRIVATE_REPO_PREFIX}{template_id}"


def template_to_response(template: Template) -> TemplateResponse:
    """Convert Template model to TemplateResponse"""
    return TemplateResponse(
        id=getattr(template, 'template_id'),
        name=getattr(template, 'name'),
        description=getattr(template, 'description'),
        category=getattr(template, 'category'),
        language=getattr(template, 'language'),
        cpu_spec=getattr(template, 'min_cpu_required'),
        memory_spec=getattr(template, 'min_memory_required'),
        is_official=getattr(template, 'is_official'),
        is_public=getattr(template, 'is_public'),
        owner_id=getattr(template.owner, 'id', None) if template.owner else None,
        repository_url=getattr(template, 'repository_url'),
        tags=getattr(template, 'tags'),
        created_at=getattr(template, 'created_at'),
        updated_at=getattr(template, 'updated_at')
    )


@router.get("/", response_model=TemplateListResponse)
async def get_templates(
    skip: int = 0,
    limit: int = 100,
    category: Optional[str] = None,
    language: Optional[str] = None,
    is_official: Optional[bool] = None,
    is_public: Optional[bool] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user_required)
):
    """Get templates with filtering - users can see official/public templates and their own private templates"""
    query = db.query(Template)

    # Apply filters
    if category:
        query = query.filter(Template.category == category)
    if language:
        query = query.filter(Template.language == language)
    if is_official is not None:
        query = query.filter(Template.is_official.is_(is_official))
    if is_public is not None:
        query = query.filter(Template.is_public.is_(is_public))

    # Permission filtering: users can see official templates, public templates, and their own private templates
    if getattr(current_user, 'role', 'user') != 'admin':
        query = query.filter(
            (Template.is_official.is_(True))
            | (Template.is_public.is_(True))
            | (Template.owner_user_id == current_user.user_id)
        )

    templates = query.offset(skip).limit(limit).all()

    return TemplateListResponse(
        templates=[template_to_response(template) for template in templates],
        total=query.count()
    )


@router.get("/{template_id}", response_model=TemplateResponse)
async def get_template(
    template_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user_required)
):
    """Get a specific template by ID"""
    template = db.query(Template).filter(Template.template_id == template_id).first()
    if not template:
        raise HTTPException(status_code=404, detail="Template not found")

    # Check permissions
    if getattr(current_user, 'role', 'user') != 'admin':
        if not getattr(template, 'is_official', False) and not getattr(template, 'is_public', False) and getattr(template, 'owner_user_id', None) != getattr(current_user, 'user_id', None):
            raise HTTPException(status_code=403, detail="Access denied")

    return template_to_response(template)


@router.post("/", response_model=TemplateResponse)
async def create_template(
    template_data: TemplateCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user_required)
):
    """Create a new template - only for private templates"""
    if template_data.is_official and getattr(current_user, 'role', 'user') != 'admin':
        raise HTTPException(status_code=403, detail="Only admins can create official templates")

    # Validate name length (2-255 characters)
    if len(template_data.name.strip()) < 2:
        raise HTTPException(status_code=400, detail="Template name must be at least 2 characters")
    if len(template_data.name.strip()) > 255:
        raise HTTPException(status_code=400, detail="Template name must be 255 characters or less")

    # Generate template ID and repository URL
    template_id = generate_template_id()
    repository_url = generate_repository_url(template_id, template_data.is_official)

    # Check for duplicate name per owner
    existing_template = db.query(Template).filter(
        Template.name == template_data.name.strip(),
        Template.owner_user_id == current_user.user_id
    ).first()
    if existing_template:
        raise HTTPException(status_code=400, detail="Template name already exists for this user")

    template = Template(
        template_id=template_id,
        name=template_data.name.strip(),
        description=template_data.description,
        category=template_data.category,
        language=template_data.language,
        min_cpu_required=template_data.cpu_spec,
        min_memory_required=template_data.memory_spec,
        is_official=template_data.is_official,
        is_public=template_data.is_public,
        owner_user_id=current_user.user_id if not template_data.is_official else None,
        repository_url=repository_url,
        tags=template_data.tags
    )

    db.add(template)
    db.commit()
    db.refresh(template)

    return template_to_response(template)


@router.put("/{template_id}", response_model=TemplateResponse)
async def update_template(
    template_id: str,
    template_data: TemplateUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user_required)
):
    """Update a template - only owner or admin can update"""
    template = db.query(Template).filter(Template.template_id == template_id).first()
    if not template:
        raise HTTPException(status_code=404, detail="Template not found")

    # Check permissions
    if getattr(current_user, 'role', 'user') != 'admin' and getattr(template, 'owner_user_id', None) != current_user.user_id:
        raise HTTPException(status_code=403, detail="Access denied")

    # Official templates can only be updated by admins
    if getattr(template, 'is_official', False) and getattr(current_user, 'role', 'user') != 'admin':
        raise HTTPException(status_code=403, detail="Only admins can update official templates")

    # Validate name if being updated
    if template_data.name is not None:
        # Validate name length (2-255 characters)
        if len(template_data.name.strip()) < 2:
            raise HTTPException(status_code=400, detail="Template name must be at least 2 characters")
        if len(template_data.name.strip()) > 255:
            raise HTTPException(status_code=400, detail="Template name must be 255 characters or less")
        
        # Check for duplicate name per owner (excluding current template)
        existing_template = db.query(Template).filter(
            Template.name == template_data.name.strip(),
            Template.owner_user_id == current_user.user_id,
            Template.template_id != template_id
        ).first()
        if existing_template:
            raise HTTPException(status_code=400, detail="A template with this name already exists")

    # Update fields
    update_data = template_data.dict(exclude_unset=True)
    for field, value in update_data.items():
        if field == 'name':
            setattr(template, field, value.strip())
        elif field == 'cpu_spec':
            setattr(template, 'min_cpu_required', value)
        elif field == 'memory_spec':
            setattr(template, 'min_memory_required', value)
        else:
            setattr(template, field, value)

    # Update the updated_at field using setattr to avoid SQLAlchemy column assignment issues
    setattr(template, 'updated_at', datetime.now(timezone.utc))
    
    try:
        db.commit()
        db.refresh(template)
    except Exception as e:
        db.rollback()
        # Check if it's a duplicate name error from database constraint
        error_str = str(e).lower()
        if "unique_template_name_per_owner" in error_str or "duplicate" in error_str:
            raise HTTPException(
                status_code=400,
                detail="A template with this name already exists"
            )
        else:
            # Re-raise other database errors
            raise HTTPException(
                status_code=500,
                detail="Failed to update template. Please try again."
            )

    return template_to_response(template)


@router.delete("/{template_id}")
async def delete_template(
    template_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user_required)
):
    """Delete a template - only owner or admin can delete. Allow deletion even if sandboxes reference it."""
    template = db.query(Template).filter(Template.template_id == template_id).first()
    if not template:
        raise HTTPException(status_code=404, detail="Template not found")

    # Check permissions
    if getattr(current_user, 'role', 'user') != 'admin' and getattr(template, 'owner_user_id', None) != current_user.user_id:
        raise HTTPException(status_code=403, detail="Access denied")

    # Official and public templates can only be deleted by admins
    if getattr(template, 'is_official', False) and getattr(current_user, 'role', 'user') != 'admin':
        raise HTTPException(status_code=403, detail="Only admins can delete official templates")
    
    if getattr(template, 'is_public', False) and getattr(current_user, 'role', 'user') != 'admin':
        raise HTTPException(status_code=403, detail="Only admins can delete public templates")

    # No longer check for sandboxes using this template. Allow deletion.

    # Send notification if admin is deleting
    if getattr(current_user, 'role', 'user') == 'admin' and getattr(template, 'owner_user_id', None):
        notification = Notification(
            user_id=template.owner_user_id,
            title="Template Deleted by Admin",
            message=f"Your template '{template.name}' has been deleted by an administrator.",
            type="warning",
            related_entity_type="template",
            related_entity_id=template_id
        )
        db.add(notification)

    db.delete(template)
    db.commit()

    return {"message": "Template deleted successfully"}


# Admin endpoints
@router.get("/admin/all", response_model=TemplateListResponse)
async def admin_get_all_templates(
    skip: int = 0,
    limit: int = 100,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user)
):
    """Admin endpoint to get all templates"""
    query = db.query(Template)
    templates = query.offset(skip).limit(limit).all()

    return TemplateListResponse(
        templates=[template_to_response(template) for template in templates],
        total=query.count()
    )


@router.post("/admin/{template_id}/make-official")
async def admin_make_template_official(
    template_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user)
):
    """Admin endpoint to make a template official"""
    template = db.query(Template).filter(Template.template_id == template_id).first()
    if not template:
        raise HTTPException(status_code=404, detail="Template not found")

    if getattr(template, 'is_official', False):
        raise HTTPException(status_code=400, detail="Template is already official")

    setattr(template, 'is_official', True)
    setattr(template, 'owner_user_id', None)  # Official templates have no owner
    setattr(template, 'updated_at', datetime.now(timezone.utc))

    # Send notification to original owner if exists
    if getattr(template, 'owner_user_id', None):
        notification = Notification(
            user_id=template.owner_user_id,
            title="Template Made Official",
            message=f"Your template '{template.name}' has been made official by an administrator.",
            type="info",
            related_entity_type="template",
            related_entity_id=template_id
        )
        db.add(notification)

    db.commit()

    return {"message": f"Template '{template.name}' made official successfully"}


@router.post("/admin/{template_id}/make-public")
async def admin_make_template_public(
    template_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user)
):
    """Admin endpoint to make a template public"""
    template = db.query(Template).filter(Template.template_id == template_id).first()
    if not template:
        raise HTTPException(status_code=404, detail="Template not found")

    if getattr(template, 'is_public', False):
        raise HTTPException(status_code=400, detail="Template is already public")

    setattr(template, 'is_public', True)
    setattr(template, 'updated_at', datetime.now(timezone.utc))

    # Send notification to owner
    if getattr(template, 'owner_user_id', None):
        notification = Notification(
            user_id=template.owner_user_id,
            title="Template Made Public",
            message=f"Your template '{template.name}' has been made public by an administrator.",
            type="info",
            related_entity_type="template",
            related_entity_id=template_id
        )
        db.add(notification)

    db.commit()

    return {"message": f"Template '{template.name}' made public successfully"} 

File: go-backend/internal/api/templates.go
package api

import (
	"encoding/json"
	"net/http"

	"github.com/gin-gonic/gin"

	"scalebox-backend/internal/models"
	"scalebox-backend/internal/utils"
)

func (s *Server) handleListTemplates(c *gin.Context) {
	var templates []models.Template

	// Get public and official templates
	query := s.db.DB.Where("is_public = ? OR is_official = ?", true, true)

	// Add category filter if provided
	if category := c.Query("category"); category != "" {
		query = query.Where("category = ?", category)
	}

	// Add language filter if provided
	if language := c.Query("language"); language != "" {
		query = query.Where("language = ?", language)
	}

	if err := query.Find(&templates).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch templates"})
		return
	}

	var response []gin.H
	for _, template := range templates {
		// Parse tags back to array for response
		var tags []string
		if template.Tags != nil {
			json.Unmarshal([]byte(*template.Tags), &tags)
		}

		response = append(response, gin.H{
			"template_id":    template.TemplateID,
			"name":           template.Name,
			"description":    template.Description,
			"category":       template.Category,
			"language":       template.Language,
			"cpu_spec":       template.MinCPURequired,
			"memory_spec":    template.MinMemoryRequired,
			"is_official":    template.IsOfficial,
			"is_public":      template.IsPublic,
			"repository_url": template.RepositoryURL,
			"tags":           tags,
			"created_at":     template.CreatedAt,
			"updated_at":     template.UpdatedAt,
		})
	}

	c.JSON(http.StatusOK, gin.H{
		"templates": response,
		"total":     len(response),
	})
}

func (s *Server) handleGetTemplate(c *gin.Context) {
	templateID := c.Param("template_id")

	var template models.Template
	if err := s.db.DB.Where("template_id = ?", templateID).First(&template).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Template not found"})
		return
	}

	// Check if template is accessible (public, official, or owned by user)
	if !template.IsPublic && !template.IsOfficial {
		user := c.MustGet("user").(models.User)
		if template.OwnerUserID == nil || *template.OwnerUserID != user.UserID {
			c.JSON(http.StatusForbidden, gin.H{"error": "Access denied"})
			return
		}
	}

	// Parse tags back to array for response
	var tags []string
	if template.Tags != nil {
		json.Unmarshal([]byte(*template.Tags), &tags)
	}

	response := gin.H{
		"template_id":    template.TemplateID,
		"name":           template.Name,
		"description":    template.Description,
		"category":       template.Category,
		"language":       template.Language,
		"cpu_spec":       template.MinCPURequired,
		"memory_spec":    template.MinMemoryRequired,
		"is_official":    template.IsOfficial,
		"is_public":      template.IsPublic,
		"repository_url": template.RepositoryURL,
		"tags":           tags,
		"created_at":     template.CreatedAt,
		"updated_at":     template.UpdatedAt,
	}

	c.JSON(http.StatusOK, response)
}

type CreateTemplateRequest struct {
	Name              string   `json:"name" binding:"required"`
	Description       string   `json:"description"`
	Category          string   `json:"category" binding:"required"`
	Language          string   `json:"language" binding:"required"`
	MinCPURequired    float64  `json:"min_cpu_required" binding:"required"`
	MinMemoryRequired float64  `json:"min_memory_required" binding:"required"`
	IsOfficial        bool     `json:"is_official"`
	IsPublic          bool     `json:"is_public"`
	RepositoryURL     string   `json:"repository_url"`
	Tags              []string `json:"tags"`
}

type UpdateTemplateRequest struct {
	Name              string   `json:"name"`
	Description       string   `json:"description"`
	Category          string   `json:"category"`
	Language          string   `json:"language"`
	MinCPURequired    float64  `json:"min_cpu_required"`
	MinMemoryRequired float64  `json:"min_memory_required"`
	IsPublic          bool     `json:"is_public"`
	RepositoryURL     string   `json:"repository_url"`
	Tags              []string `json:"tags"`
}

func (s *Server) handleCreateTemplate(c *gin.Context) {
	user := c.MustGet("user").(models.User)

	var req CreateTemplateRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	// Convert tags to JSON string
	var tagsJSON *string
	if req.Tags != nil {
		tagsBytes, err := json.Marshal(req.Tags)
		if err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid tags format"})
			return
		}
		tagsStr := string(tagsBytes)
		tagsJSON = &tagsStr
	}

	template := models.Template{
		TemplateID:        utils.GenerateTemplateID(),
		Name:              req.Name,
		Description:       &req.Description,
		Category:          req.Category,
		Language:          req.Language,
		MinCPURequired:    req.MinCPURequired,
		MinMemoryRequired: req.MinMemoryRequired,
		IsOfficial:        req.IsOfficial,
		IsPublic:          req.IsPublic,
		RepositoryURL:     req.RepositoryURL,
		Tags:              tagsJSON,
		OwnerUserID:       &user.UserID,
	}

	if err := s.db.DB.Create(&template).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create template"})
		return
	}

	// Parse tags back to array for response
	var tags []string
	if template.Tags != nil {
		json.Unmarshal([]byte(*template.Tags), &tags)
	}

	response := gin.H{
		"template_id":         template.TemplateID,
		"name":                template.Name,
		"description":         template.Description,
		"category":            template.Category,
		"language":            template.Language,
		"min_cpu_required":    template.MinCPURequired,
		"min_memory_required": template.MinMemoryRequired,
		"is_official":         template.IsOfficial,
		"is_public":           template.IsPublic,
		"repository_url":      template.RepositoryURL,
		"tags":                tags,
		"owner_id":            template.OwnerUserID,
		"created_at":          template.CreatedAt,
		"updated_at":          template.UpdatedAt,
	}

	c.JSON(http.StatusCreated, response)
}

func (s *Server) handleUpdateTemplate(c *gin.Context) {
	user := c.MustGet("user").(models.User)
	templateID := c.Param("template_id")

	var template models.Template
	if err := s.db.DB.Where("template_id = ?", templateID).First(&template).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Template not found"})
		return
	}

	// Check if user can edit this template
	if template.IsOfficial || (template.IsPublic && (template.OwnerUserID == nil || *template.OwnerUserID != user.UserID)) {
		c.JSON(http.StatusForbidden, gin.H{"error": "Cannot edit this template"})
		return
	}

	var req UpdateTemplateRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	// Update fields
	if req.Name != "" {
		template.Name = req.Name
	}
	if req.Description != "" {
		template.Description = &req.Description
	}
	if req.Category != "" {
		template.Category = req.Category
	}
	if req.Language != "" {
		template.Language = req.Language
	}
	if req.MinCPURequired > 0 {
		template.MinCPURequired = req.MinCPURequired
	}
	if req.MinMemoryRequired > 0 {
		template.MinMemoryRequired = req.MinMemoryRequired
	}
	template.IsPublic = req.IsPublic
	if req.RepositoryURL != "" {
		template.RepositoryURL = req.RepositoryURL
	}
	if req.Tags != nil {
		// Convert tags to JSON string
		tagsBytes, err := json.Marshal(req.Tags)
		if err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid tags format"})
			return
		}
		tagsStr := string(tagsBytes)
		template.Tags = &tagsStr
	}

	if err := s.db.DB.Save(&template).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update template"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Template updated successfully"})
}

func (s *Server) handleDeleteTemplate(c *gin.Context) {
	user := c.MustGet("user").(models.User)
	templateID := c.Param("template_id")

	var template models.Template
	if err := s.db.DB.Where("template_id = ?", templateID).First(&template).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Template not found"})
		return
	}

	// Check if user can delete this template
	if template.IsOfficial || (template.IsPublic && (template.OwnerUserID == nil || *template.OwnerUserID != user.UserID)) {
		c.JSON(http.StatusForbidden, gin.H{"error": "Cannot delete this template"})
		return
	}

	if err := s.db.DB.Delete(&template).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to delete template"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Template deleted successfully"})
}
